### PR TITLE
Fix issues found with flint

### PR DESCRIPTION
**Pull request by evincarofautumn**
[`flint`](https://github.com/facebookarchive/flint) is a C++ linter that is (was?) used at Facebook. These changes can be cherry-picked if not all of them are desirable.

* Remove unquoted <code>\`</code> and `^L` because compilers are not required to support them.

* Remove duplicate `#include` lines.

* Use `__attribute__ ((__foo__))` instead of `__attribute__ ((foo))` so that we can “use them in header files without being concerned about a possible macro of the same name” ([GCC manual](https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html#Attribute-Syntax)).

* Avoid `strncpy` because [somebody said so](http://meyering.net/crusade-to-eliminate-strncpy/).

### CODE REVIEW

File: mono/metadata/appdomain.c [hunk 1]
```diff
@@ -44,7 +44,6 @@
 #include <mono/metadata/threads.h>
 #include <mono/metadata/threadpool.h>
 #include <mono/metadata/tabledefs.h>
-#include <mono/metadata/gc-internals.h>
 #include <mono/metadata/mono-gc.h>
```
Comment (vargaz) [inline mono/metadata/appdomain.c]: Some people have the opinion that files should include what they use and not depend on it being included from other include files.

File: mono/arch/s390x/s390x-codegen.h [hunk 2]
```diff
@@ -268,7 +268,7 @@ typedef struct {
 	char	xx : 4;
 	char	r3 : 4;
 	char	r4 : 4;
-} __attribute__ ((packed)) RRD_Format;
```
Comment (vargaz) [inline mono/arch/s390x/s390x-codegen.h]: What is the problem with the non-underscore attributes ?

File: mono/arch/s390x/s390x-codegen.h [hunk 3]
```diff
@@ -268,7 +268,7 @@ typedef struct {
...
+} __attribute__ ((__packed__)) RRD_Format;
```
Comment (vargaz) [inline mono/arch/s390x/s390x-codegen.h]: Never seen attributes used like this anywhere.

File: mono/metadata/appdomain.c (hunk 1 above)
Comment (evincarofautumn) [inline mono/metadata/appdomain.c]: True, and Flint does have an “Include What You Use” mode. However, I only removed includes when there was an identical, unconditional include in the same file.

File: mono/arch/s390x/s390x-codegen.h (hunk 3 above)
Comment (evincarofautumn) [inline mono/arch/s390x/s390x-codegen.h]: Neither had I. It’s not a very important change, but it doesn’t hurt.

Comment (alexrp) [inline mono/arch/s390x/s390x-codegen.h]: But what is the reason for this change? It's GCC-specific syntax anyway, so why make it look even uglier?

Comment (evincarofautumn) [inline mono/arch/s390x/s390x-codegen.h]: I just PR’d exactly what Flint suggested, to see which changes were wanted.

Their reasoning is that identifiers containing `__` are reserved, so the underscore-adorned attribute should always work even in the presence of macros such as `#define error abort` or `#define printf g_printf`. As far as I can tell, the embedding API is currently the only place this might matter for Mono.

Clang also appears to accept this syntax—I was under the impression that their policy is to support all GNU attributes and extensions, but I can’t seem to find a citation for that.

File: mono/metadata/appdomain.c (hunk 1 above)
Comment (vargaz) [inline mono/metadata/appdomain.c]: Its ok then, didn't notice the double include.

File: mono/metadata/appdomain.c [hunk 4]
```diff
@@ -131,9 +131,9 @@ get_shadow_assembly_location_base (MonoDomain *domain, MonoError *error);
 static MonoLoadFunc load_function = NULL;
 
 /* Lazy class loading functions */
-static GENERATE_GET_CLASS_WITH_CACHE (assembly, System.Reflection, Assembly);
+static GENERATE_GET_CLASS_WITH_CACHE (assembly, System.Reflection, "Assembly");
```
Comment (vargaz) [inline mono/metadata/appdomain.c]: This looks ok, but what problem it is trying to avoid ?

Comment (evincarofautumn) [inline mono/metadata/appdomain.c]: There were a couple of places we used e.g. ``IEnumerable`1`` unquoted. These can just be strings because we immediately quoted them with `#name` anyway.

Comment (vargaz) [inline mono/metadata/appdomain.c]: We should probably use strings for the namespace as well.
